File: retail/retail_app/report/tax_report/tax_report.js
frappe.query_reports["Tax Report"] = {
	onload: function (report) {
		include_checkbox_filters(report);
		set_party_filter_options(report);
		add_tax_report_print_buttons(report, "Landscape");
	},
	refresh: function (report) {
		add_tax_report_print_buttons(report, "Landscape");
	},
	filters: [
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			default: frappe.defaults.get_user_default("Company"),
			reqd: 1,
		},
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
			default: frappe.datetime.month_start(),
			reqd: 1,
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
			default: frappe.datetime.get_today(),
			reqd: 1,
		},
		{
			fieldname: "cost_center",
			label: __("Cost Center"),
			fieldtype: "Link",
			options: "Cost Center",
			get_query: function () {
				return {
					filters: {
						company: frappe.query_report.get_filter_value("company"),
						is_group: 0,
					},
				};
			},
		},
		{
			fieldname: "transaction_type",
			label: __("Transaction Type"),
			fieldtype: "Select",
			options: "\nPOS Sales\nPOS Sales Return\nSales Invoice\nSales Return\nPurchase Invoice\nPurchase Return\nExpense\nTax Adjustment",
		},
		{
			fieldname: "party_type",
			label: __("Party Type"),
			fieldtype: "Select",
			options: ["", "Customer", "Supplier"],
			on_change: function (report) {
				set_party_filter_options(report);
				Promise.resolve(report.set_filter_value("party", "")).then(() => report.refresh(true));
			},
		},
		{
			fieldname: "party",
			label: __("Party"),
			fieldtype: "Link",
			options: "Customer",
			get_query: function () {
				const party_type = frappe.query_report.get_filter_value("party_type") || "Customer";
				const filters = {};
				if (party_type === "Customer") {
					filters.disabled = 0;
				} else if (party_type === "Supplier") {
					filters.disabled = 0;
				}
				return { doctype: party_type, filters };
			},
		},
		{
			fieldname: "tax_account",
			label: __("Tax Account"),
			fieldtype: "Link",
			options: "Account",
			get_query: function () {
				return {
					filters: {
						company: frappe.query_report.get_filter_value("company"),
						account_type: "Tax",
						is_group: 0,
					},
				};
			},
		},
		{
			fieldname: "show_zero_vat_rows",
			label: __("Show Zero Tax Rows"),
			fieldtype: "Check",
			default: 0,
			on_change: function (report) {
				report.refresh(true);
			},
		},
		{
			fieldname: "show_summary_rows",
			label: __("Show Summary Rows"),
			fieldtype: "Check",
			default: 1,
			on_change: function (report) {
				report.refresh(true);
			},
		},
	],
	formatter: function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		if (data && data.is_group) {
			value = $(`<span>${value || ""}</span>`).css("font-weight", "600").wrap("<p></p>").parent().html();
		}

		return value;
	},
};

function set_party_filter_options(report) {
	const party_type = report.get_filter_value("party_type") || "Customer";
	const party_filter = report.get_filter("party");

	if (!party_filter) {
		return;
	}

	party_filter.df.options = party_type;
	party_filter.$input && party_filter.$input.attr("data-target", party_type);
	if (party_filter.$input) {
		party_filter.$input.cache = {};
	}
	if (party_filter.awesomplete) {
		party_filter.awesomplete.list = [];
	}
	party_filter.refresh();
}

function include_checkbox_filters(report) {
	if (report._tax_report_checkbox_filters_included) {
		return;
	}

	const original_get_filter_values = report.get_filter_values.bind(report);
	report.get_filter_values = function (raise) {
		const values = original_get_filter_values(raise);
		values.show_zero_vat_rows = cint(report.get_filter_value("show_zero_vat_rows", false));
		values.show_summary_rows = cint(report.get_filter_value("show_summary_rows", false));
		return values;
	};
	report._tax_report_checkbox_filters_included = true;
}

function add_tax_report_print_buttons(report, orientation) {
	if (report._custom_tax_report_print_buttons_added) {
		return;
	}

	const add_buttons = () => {
		if (!report.page) {
			setTimeout(add_buttons, 300);
			return;
		}

		const show_print_settings = (callback) => {
			frappe.ui.get_print_settings(
				false,
				(print_settings) => {
					print_settings.orientation = orientation;
					print_settings.include_filters = 0;
					print_settings.print_format = null;
					print_settings.report = null;
					print_settings.columns = [];
					print_settings.pick_columns = 0;
					callback(print_settings);
				},
				report.report_doc.letter_head,
				false,
				false
			);
		};

		const print_action = () => show_print_settings((settings) => report.print_report(settings));
		const pdf_action = () => show_print_settings((settings) => report.pdf_report(settings));

		report.page.add_inner_button(__("Custom Print"), print_action);
		report.page.add_inner_button(__("Custom PDF"), pdf_action);
		report.page.add_menu_item(__("Custom Print"), print_action);
		report.page.add_menu_item(__("Custom PDF"), pdf_action);
		report._custom_tax_report_print_buttons_added = true;
	};

	setTimeout(add_buttons, 300);
}

File: retail/retail_app/report/tax_report/tax_report.py
import frappe
from frappe import _
from frappe.utils import cint, flt, getdate


def execute(filters=None):
	filters = frappe._dict(filters or {})
	validate_filters(filters)

	rows = []
	rows.extend(get_sales_invoice_rows(filters))
	rows.extend(get_purchase_invoice_rows(filters))
	rows.extend(get_journal_entry_rows(filters))

	rows = apply_common_filters(rows, filters)
	rows.sort(key=lambda row: (row.transaction_type, getdate(row.date), row.voucher_no or ""))
	add_reference_numbers(rows)

	if cint(filters.get("show_summary_rows", 1)):
		rows = add_summary_rows(rows)

	return get_columns(), rows


def validate_filters(filters):
	if not filters.get("company"):
		frappe.throw(_("Company is required"))
	if not filters.get("from_date") or not filters.get("to_date"):
		frappe.throw(_("From Date and To Date are required"))
	if getdate(filters.from_date) > getdate(filters.to_date):
		frappe.throw(_("From Date cannot be after To Date"))


def get_uae_vat_accounts(company):
	accounts = frappe.get_all("UAE VAT Account", filters={"parent": company}, pluck="account")
	if accounts:
		return accounts

	return frappe.get_all(
		"Account",
		filters={"company": company, "account_type": "Tax", "is_group": 0},
		pluck="name",
	)


def get_sales_invoice_rows(filters):
	tax_accounts = get_uae_vat_accounts(filters.company)
	if not tax_accounts:
		return []

	conditions = [
		"si.docstatus = 1",
		"si.company = %(company)s",
		"si.posting_date between %(from_date)s and %(to_date)s",
	]

	if filters.get("cost_center"):
		conditions.append(
			"""(
				si.cost_center = %(cost_center)s
				or exists (
					select 1 from `tabSales Invoice Item` sii
					where sii.parent = si.name and sii.cost_center = %(cost_center)s
				)
				or exists (
					select 1 from `tabSales Taxes and Charges` stc
					where stc.parent = si.name and stc.cost_center = %(cost_center)s
				)
			)"""
		)

	rows = frappe.db.sql(
		f"""
		select
			si.name as voucher_no,
			si.name as invoice_no,
			si.posting_date as date,
			si.customer as party,
			coalesce(si.customer_name, si.customer) as party_name,
			si.tax_id as trn,
			coalesce(
				si.cost_center,
				(select max(sii.cost_center) from `tabSales Invoice Item` sii where sii.parent = si.name),
				(select max(stc.cost_center) from `tabSales Taxes and Charges` stc where stc.parent = si.name)
			) as cost_center,
			si.is_pos,
			si.is_return,
			abs(si.base_net_total) as net_total,
			abs(coalesce(si.base_discount_amount, 0)) as discount_amount,
			abs((
				select sum(coalesce(stc.base_tax_amount_after_discount_amount, stc.base_tax_amount, stc.tax_amount, 0))
				from `tabSales Taxes and Charges` stc
				where stc.parent = si.name and stc.account_head in %(tax_accounts)s
			)) as vat_amount,
			abs(si.base_grand_total) as gross_amount,
			(
				select group_concat(distinct stc.account_head order by stc.account_head separator ', ')
				from `tabSales Taxes and Charges` stc
				where stc.parent = si.name and stc.account_head in %(tax_accounts)s
			) as tax_account
		from `tabSales Invoice` si
		where {" and ".join(conditions)}
			and exists (
				select 1 from `tabSales Taxes and Charges` stc
				where stc.parent = si.name and stc.account_head in %(tax_accounts)s
			)
		""",
		{
			"company": filters.company,
			"from_date": filters.from_date,
			"to_date": filters.to_date,
			"cost_center": filters.get("cost_center"),
			"tax_accounts": tuple(tax_accounts),
		},
		as_dict=True,
	)

	data = []
	for row in rows:
		is_return = bool(row.is_return)
		row.transaction_type = get_sales_transaction_type(row)
		sign = -1 if is_return else 1
		row.total = sign * flt(row.net_total)
		row.discount_amount = sign * flt(row.discount_amount)
		row.taxable_amount = sign * flt(row.net_total) if flt(row.vat_amount) else 0
		row.non_taxable_amount = 0 if flt(row.vat_amount) else sign * flt(row.net_total)
		row.output_tax = sign * flt(row.vat_amount)
		row.input_tax = 0
		row.net_amount = sign * flt(row.gross_amount)
		row.tax_payable = None
		row.party_type = "Customer"
		row.doctype = "Sales Invoice"
		data.append(row)

	return data


def get_sales_transaction_type(row):
	if row.is_pos and row.is_return:
		return _("POS Sales Return")
	if row.is_pos:
		return _("POS Sales")
	if row.is_return:
		return _("Sales Return")
	return _("Sales Invoice")


def get_purchase_invoice_rows(filters):
	tax_accounts = get_uae_vat_accounts(filters.company)
	if not tax_accounts:
		return []

	conditions = [
		"pi.docstatus = 1",
		"pi.company = %(company)s",
		"pi.posting_date between %(from_date)s and %(to_date)s",
	]

	if filters.get("cost_center"):
		conditions.append(
			"""(
				pi.cost_center = %(cost_center)s
				or exists (
					select 1 from `tabPurchase Invoice Item` pii
					where pii.parent = pi.name and pii.cost_center = %(cost_center)s
				)
				or exists (
					select 1 from `tabPurchase Taxes and Charges` ptc
					where ptc.parent = pi.name and ptc.cost_center = %(cost_center)s
				)
			)"""
		)

	rows = frappe.db.sql(
		f"""
		select
			pi.name as voucher_no,
			coalesce(nullif(pi.bill_no, ''), pi.name) as invoice_no,
			pi.posting_date as date,
			pi.supplier as party,
			coalesce(pi.supplier_name, pi.supplier) as party_name,
			pi.tax_id as trn,
			coalesce(
				pi.cost_center,
				(select max(pii.cost_center) from `tabPurchase Invoice Item` pii where pii.parent = pi.name),
				(select max(ptc.cost_center) from `tabPurchase Taxes and Charges` ptc where ptc.parent = pi.name)
			) as cost_center,
			pi.is_return,
			abs(pi.base_net_total) as net_total,
			abs(coalesce(pi.base_discount_amount, 0)) as discount_amount,
			abs((
				select sum(coalesce(ptc.base_tax_amount_after_discount_amount, ptc.base_tax_amount, ptc.tax_amount, 0))
				from `tabPurchase Taxes and Charges` ptc
				where ptc.parent = pi.name and ptc.account_head in %(tax_accounts)s
			)) as vat_amount,
			abs(pi.base_grand_total) as gross_amount,
			(
				select group_concat(distinct ptc.account_head order by ptc.account_head separator ', ')
				from `tabPurchase Taxes and Charges` ptc
				where ptc.parent = pi.name and ptc.account_head in %(tax_accounts)s
			) as tax_account,
			(
				select count(*)
				from `tabPurchase Invoice Item` pii
				inner join `tabItem` item on item.name = pii.item_code
				where pii.parent = pi.name and item.is_stock_item = 1
			) as stock_item_count
		from `tabPurchase Invoice` pi
		where {" and ".join(conditions)}
			and exists (
				select 1 from `tabPurchase Taxes and Charges` ptc
				where ptc.parent = pi.name and ptc.account_head in %(tax_accounts)s
			)
		""",
		{
			"company": filters.company,
			"from_date": filters.from_date,
			"to_date": filters.to_date,
			"cost_center": filters.get("cost_center"),
			"tax_accounts": tuple(tax_accounts),
		},
		as_dict=True,
	)

	data = []
	for row in rows:
		is_return = bool(row.is_return)
		row.transaction_type = _("Purchase Return") if is_return else get_purchase_transaction_type(row)
		sign = -1 if is_return else 1
		row.total = sign * flt(row.net_total)
		row.discount_amount = sign * flt(row.discount_amount)
		row.taxable_amount = sign * flt(row.net_total) if flt(row.vat_amount) else 0
		row.non_taxable_amount = 0 if flt(row.vat_amount) else sign * flt(row.net_total)
		row.output_tax = 0
		row.input_tax = sign * flt(row.vat_amount)
		row.net_amount = sign * flt(row.gross_amount)
		row.tax_payable = None
		row.party_type = "Supplier"
		row.doctype = "Purchase Invoice"
		data.append(row)

	return data


def get_purchase_transaction_type(row):
	return _("Purchase Invoice") if flt(row.stock_item_count) else _("Expense")


def get_journal_entry_rows(filters):
	tax_accounts = get_uae_vat_accounts(filters.company)
	if not tax_accounts:
		return []

	conditions = [
		"je.docstatus = 1",
		"je.company = %(company)s",
		"je.posting_date between %(from_date)s and %(to_date)s",
		"jea.account in %(tax_accounts)s",
	]

	if filters.get("cost_center"):
		conditions.append("jea.cost_center = %(cost_center)s")

	rows = frappe.db.sql(
		f"""
		select
			je.name as voucher_no,
			coalesce(nullif(je.bill_no, ''), je.name) as invoice_no,
			je.posting_date as date,
			jea.party_type,
			jea.party,
			coalesce(jea.cost_center, '') as cost_center,
			jea.account as tax_account,
			sum(jea.debit) as input_tax,
			sum(jea.credit) as output_tax,
			je.user_remark as remarks
		from `tabJournal Entry` je
		inner join `tabJournal Entry Account` jea on jea.parent = je.name
		where {" and ".join(conditions)}
		group by je.name, jea.account, jea.cost_center, jea.party_type, jea.party
		""",
		{
			"company": filters.company,
			"from_date": filters.from_date,
			"to_date": filters.to_date,
			"cost_center": filters.get("cost_center"),
			"tax_accounts": tuple(tax_accounts),
		},
		as_dict=True,
	)

	for row in rows:
		row.transaction_type = _("Tax Adjustment")
		row.party_name = get_party_name(row.party_type, row.party)
		row.trn = get_party_trn(row.party_type, row.party)
		row.total = 0
		row.discount_amount = 0
		row.taxable_amount = 0
		row.non_taxable_amount = 0
		row.net_amount = 0
		row.tax_payable = None
		row.doctype = "Journal Entry"

	return rows


def get_party_name(party_type, party):
	if not party_type or not party:
		return ""
	if party_type == "Customer":
		return frappe.get_cached_value("Customer", party, "customer_name") or party
	if party_type == "Supplier":
		return frappe.get_cached_value("Supplier", party, "supplier_name") or party
	return party


def get_party_trn(party_type, party):
	if not party_type or not party:
		return ""
	if party_type in ("Customer", "Supplier"):
		return frappe.get_cached_value(party_type, party, "tax_id") or ""
	return ""


def apply_common_filters(rows, filters):
	filtered = []
	for row in rows:
		if filters.get("transaction_type") and row.transaction_type != filters.transaction_type:
			continue
		if filters.get("tax_account") and filters.tax_account not in (row.tax_account or ""):
			continue
		if filters.get("party_type") and row.get("party_type") != filters.party_type:
			continue
		if filters.get("party") and row.get("party") != filters.party:
			continue
		if not cint(filters.get("show_zero_vat_rows")) and not (flt(row.output_tax) or flt(row.input_tax)):
			continue
		filtered.append(row)
	return filtered


def add_reference_numbers(rows):
	for index, row in enumerate(rows, 1):
		row.ref_no = index


def add_summary_rows(rows):
	grouped_rows = []
	current_type = None
	totals = frappe._dict()
	grand_totals = frappe._dict()

	for row in rows:
		if current_type and row.transaction_type != current_type:
			grouped_rows.append(make_total_row(current_type, totals))
			totals = frappe._dict()

		current_type = row.transaction_type
		grouped_rows.append(row)
		add_to_totals(totals, row)
		add_to_totals(grand_totals, row)

	if current_type:
		grouped_rows.append(make_total_row(current_type, totals))
		grouped_rows.append(make_total_row(_("Grand Total"), grand_totals, is_grand_total=True))
		grouped_rows.append(make_vat_payable_row(grand_totals))

	return grouped_rows


def add_to_totals(totals, row):
	for field in (
		"total",
		"discount_amount",
		"taxable_amount",
		"non_taxable_amount",
		"output_tax",
		"input_tax",
		"net_amount",
	):
		totals[field] = flt(totals.get(field)) + flt(row.get(field))


def make_total_row(label, totals, is_grand_total=False):
	return frappe._dict(
		{
			"transaction_type": _("{0} Total").format(label) if not is_grand_total else label,
			"total": totals.get("total"),
			"discount_amount": totals.get("discount_amount"),
			"taxable_amount": totals.get("taxable_amount"),
			"non_taxable_amount": totals.get("non_taxable_amount"),
			"output_tax": totals.get("output_tax"),
			"input_tax": totals.get("input_tax"),
			"net_amount": totals.get("net_amount"),
			"tax_payable": None,
			"is_group": 1,
		}
	)


def make_vat_payable_row(totals):
	return frappe._dict(
		{
			"transaction_type": _("Tax Payable / Recoverable"),
			"tax_payable": flt(totals.get("output_tax")) - flt(totals.get("input_tax")),
			"is_group": 1,
		}
	)


def get_columns():
	return [
		{"label": _("Date"), "fieldname": "date", "fieldtype": "Date", "width": 105},
		{"label": _("Transaction Type"), "fieldname": "transaction_type", "fieldtype": "Data", "width": 150},
		{
			"label": _("Voucher No"),
			"fieldname": "voucher_no",
			"fieldtype": "Dynamic Link",
			"options": "doctype",
			"width": 180,
		},
		{"label": _("Invoice No"), "fieldname": "invoice_no", "fieldtype": "Data", "width": 140},
		{"label": _("Party Name"), "fieldname": "party_name", "fieldtype": "Data", "width": 220},
		{"label": _("Party Tax ID"), "fieldname": "trn", "fieldtype": "Data", "width": 140},
		{"label": _("Cost Center"), "fieldname": "cost_center", "fieldtype": "Link", "options": "Cost Center", "width": 140},
		{"label": _("Total"), "fieldname": "total", "fieldtype": "Currency", "width": 120},
		{"label": _("Discount"), "fieldname": "discount_amount", "fieldtype": "Currency", "width": 110},
		{"label": _("Taxable Amount"), "fieldname": "taxable_amount", "fieldtype": "Currency", "width": 130},
		{"label": _("Non Taxable Amount"), "fieldname": "non_taxable_amount", "fieldtype": "Currency", "width": 150},
		{"label": _("Output Tax"), "fieldname": "output_tax", "fieldtype": "Currency", "width": 115},
		{"label": _("Input Tax"), "fieldname": "input_tax", "fieldtype": "Currency", "width": 110},
		{"label": _("Net Amount"), "fieldname": "net_amount", "fieldtype": "Currency", "width": 125},
		{"label": _("Tax Payable / Recoverable"), "fieldname": "tax_payable", "fieldtype": "Currency", "width": 170},
		{"label": _("Tax Account"), "fieldname": "tax_account", "fieldtype": "Data", "width": 220},
		{"label": _("Party Type"), "fieldname": "party_type", "fieldtype": "Data", "width": 100},
		{"label": _("Party"), "fieldname": "party", "fieldtype": "Data", "width": 170},
		{"label": _("Remarks"), "fieldname": "remarks", "fieldtype": "Data", "width": 220},
		{"label": _("Doctype"), "fieldname": "doctype", "fieldtype": "Data", "width": 120, "hidden": 1},
	]
